cmd/main.go

```diff
@@ -195,31 +98,54 @@ func main() {
 		NSXConfig: cf,
 	}
 
-	if cf.CoeConfig.EnableVPCNetwork && commonService.NSXClient.NSXCheckVersion(nsx.VPC) {
-		log.V(1).Info("VPC mode enabled")
-		// Start controllers which only supports VPC
-		// Start subnet/subnetset controller.
+	if cf.CoeConfig.EnableVPCNetwork {
+		// Check NSX version for VPC networking mode
+		if !commonService.NSXClient.NSXCheckVersion(nsx.VPC) {
+			log.Error(nil, "VPC mode cannot be enabled if NSX version is lower than 4.1.1")
+			os.Exit(1)
+		}
+		log.Info("VPC mode is enabled")
+		// Start VPC controllers
+		if err := namespace.StartNamespaceController(mgr, commonService); err != nil {
+			log.Error(err, "failed to start Namespace controller")
+			os.Exit(1)
+		}
+		if err := vpc.StartVPCController(mgr, commonService); err != nil {
+			log.Error(err, "failed to start VPC controller")
+			os.Exit(1)
+		}
 		if err := subnet.StartSubnetController(mgr, commonService); err != nil {
+			log.Error(err, "failed to start Subnet controller")
 			os.Exit(1)
 		}
 		if err := subnetset.StartSubnetSetController(mgr, commonService); err != nil {
+			log.Error(err, "failed to start SubnetSet controller")
+			os.Exit(1)
+		}
+		if err := staticroute.StartStaticRouteController(mgr, commonService); err != nil {
+			log.Error(err, "failed to start StaticRoute controller")
+			os.Exit(1)
+		}
+		if err := subnetport.StartSubnetPortController(mgr, commonService); err != nil {
+			log.Error(err, "failed to start SubnetPort controller")
+			os.Exit(1)
+		}
+		if err := ippool.StartIPPoolController(mgr, commonService); err != nil {
+			log.Error(err, "failed to start IPPool controller")
 			os.Exit(1)
 		}
```

Comment:
is it possible that we use a interface Controller, and contains a unified Start method, each controller implement the start method. then we can maintain a List[Controller] , use a for loop to startup all the Controllers, this code seems quite duplicate, and the parameters are almost the same.

Reply:
yes, we could do it like Cleanup.

Reply:
refer https://github.com/vmware-tanzu/nsx-operator/pull/330
could close this pr.